ethexe/contracts/src/ScaleCodec.sol

```diff
@@ -0,0 +1,508 @@
+            for { let i := 0 } lt(i, data_len) { i := add(i, 1) } {
+                let v := mload(add(data_ptr, i))
+                mstore(add(dest_ptr, i), v)
+            }
+        }
+    }
+
+    function encodeBool(bool value) public pure returns (bytes memory) {
+        bytes memory result = new bytes(1);
+        encodeBoolTo(value, result, 0);
+        return result;
+    }
+
+    function encodeBoolTo(bool value, bytes memory destination, uint256 offset) internal pure {
+        if (value) {
+            destination[offset] = 0x01;
+        } else {
+            destination[offset] = 0x00;
+        }
+    }
+
+    function decodeBool(bytes memory _bytes, uint256 offset) public pure returns (bool) {
+        return _bytes[offset] == 0x01;
+    }
+
+    function encodeUint8(uint8 value) public pure returns (bytes memory) {
+        bytes memory destination = new bytes(1);
+        encodeUint8To(value, destination, 0);
+        return destination;
+    }
+
+    function encodeUint8To(uint8 value, bytes memory destination, uint256 offset) internal pure {
+        assembly {
+            let dest := add(add(destination, 0x20), offset)
+            mstore8(dest, value)
+        }
+    }
+
+    function decodeUint8(bytes memory _bytes, uint256 offset) public pure returns (uint8) {
+        return uint8(_bytes[offset]);
+    }
+
+    function encodeInt8(int8 value) public pure returns (bytes memory) {
+        return encodeUint8(uint8(value));
+    }
+
+    function encodeInt8To(int8 value, bytes memory destination, uint256 offset) internal pure {
+        encodeUint8To(uint8(value), destination, offset);
+    }
+
+    function decodeInt8(bytes memory _bytes, uint256 offset) public pure returns (int8) {
+        return int8(uint8(_bytes[offset]));
+    }
+
+    function encodeUint16(uint16 value) public pure returns (bytes memory) {
+        bytes memory result = new bytes(2);
+        encodeUint16To(value, result, 0);
+        return result;
+    }
+
+    function encodeUint16To(uint16 value, bytes memory destination, uint256 offset) internal pure {
+        assembly {
+            let dest := add(add(destination, 0x20), offset)
+            mstore8(dest, and(value, 0xff))
+            mstore8(add(dest, 1), shr(0x08, value))
+        }
+    }
+
+    function decodeUint16(bytes memory _bytes, uint256 offset) public pure returns (uint16) {
+        return uint16(bytesToUint(_bytes, 2, offset));
+    }
+
+    function encodeInt16(int16 value) public pure returns (bytes memory) {
+        return encodeUint16(uint16(value));
+    }
+
+    function encodeInt16To(int16 value, bytes memory destination, uint256 offset) internal pure {
+        encodeUint16To(uint16(value), destination, offset);
+    }
+
+    function decodeInt16(bytes memory _bytes, uint256 offset) public pure returns (int16) {
+        return int16(decodeUint16(_bytes, offset));
+    }
+
+    function encodeUint32(uint32 value) public pure returns (bytes memory) {
```

Comment:
do you always need `bytes`?
seems that uint32 value has zero cast to bytes4 - so it could be also useful for future usage?

Reply:
Actually `encodeUint32To` function is supposed to be used most of the time (at least in contracts generated by sails). And here I belief `bytes` is better cuz you probably need to combine those bytes with some others and you can't do it if you have type `bytes4`. And even if you don't need to combine them you need to pass those bytes to the `sendMessage` function which takes `bytes` type